Add release notes for mining changes

sdaftuar opened the pull request.

sdaftuar committed:
Add release notes for mining changes

luke-jr commented on doc/release-notes.md:
```diff
@@ -146,6 +146,41 @@ HD wallets are incompatible with older versions of Bitcoin Core.
+
+The command line option `-blockmaxsize` remains an option to specify the
+maximum number of serialized bytes in a generated block; however, use of that
+option is discouraged in favor of a a new command line option,
```
No, it should not be discouraged (yet). Miners are encouraged to set blockmaxsize to 600000-750000 (the latter being the default).

luke-jr commented:
This is missing any mention that miners should not produce blocks larger than 1 MB until the network is known to be capable of handling it. This is critical for miners to understand before BIP 141 activates.

sipa commented:
@luke-jr Do you know of a way to formulate that in a measurable way? How are miners expected to judge "the network known to be capable of handling it"?

luke-jr commented:
When most miners are comfortable shutting off centralised relay networks and removing explicit outside addnode/connect to their nodes, that seems a reasonable indication that the network is ready.

sipa commented:
I doubt they would ever be confortable doing that, even with only empty
blocks.

luke-jr commented:
Miners worked fine without centralised networks before. It's not like it's unreasonable.

sipa commented:
@luke-jr What I think the right criterion is whether the additional propagation delay from a larger block size would not affect (or only in a negligable amount) the economics of running things like relay network, SPV mining, ... etc. And it may well be the case that compact blocks, better mempool logic, validation caching, ... bring us closer to that, as they make relay times less dependent on the block size, and more on validation time (for which block weight is probably a better measure, if UTXO growth is taken into account).

I don't think release notes are the place for this kind of discussion, though. Are you ok with just removing the language about discouraging/encouraging how people use these options, and simply state their existence?

Then we can see what effect the changes in the 0.13 release have on the network, and reassess whether any recommendations are needed in the first release that actually does a block size increase (through segwit)?

sdaftuar commented:
Updated to address comments.  

I added a brief section on segwit to correct any confusion that somehow 0.13.0 supports segwit on mainnet.  We should consider expanding that section to include more helpful notes on how to use 0.13.0 on testnet.

luke-jr commented:
I'm fine with deferring any recommendations until segwit has been released for mainnet.

luke-jr commented on doc/release-notes.md:
```diff
@@ -146,6 +146,68 @@ HD wallets are incompatible with older versions of Bitcoin Core.
+activation, these calculations would be expected to differ.
+
+For optimal runtime performance, miners using this release should specify
+`-blockmaxweight` on the command line, and not specify `-blockmaxsize`.
```
NACK.

sipa replied:
Can you suggest an alternative here? I think we should somehow convey that there is a performance penalty to using -blockmaxsize.

luke-jr commented on doc/release-notes.md:
```diff
@@ -146,6 +146,68 @@ HD wallets are incompatible with older versions of Bitcoin Core.
+Additionally (or only) specifying `-blockmaxsize`, or relying on default
+settings for both, may result in performance degradation, as the logic to
+support `-blockmaxsize` performs additional computation to ensure that
+constraint is met.  (Note that for mainnet, in this release, the equivalent
```
Why is there performance degradation? That was not the case a few months ago, and we probably shouldn't release if it's noteworthy here.

sipa replied:
Because we sort based on fee per weight; checking whether the size constraint is met is an additional check that is not otherwise needed. @sdaftuar do you have any numbers?

Furthermore, it makes fee optimization for miners a non-linear optimization problem (and implementing that would make fee estimation much more complicated as well, because wallets would need to know whether weight or size will be the constraining factor).

luke-jr replied:
So if ordering by fee-per-bytes is a performance issue, just order by fee-per-weight and close the block out when the size limit is hit. The overhead for that check should be negligible.

As this stands, it is de facto pushing miners toward a configuration that ought to be discouraged when segwit activates.

sdaftuar replied:
I don't have any numbers, and as far as I know no one has attempted to measure this.

I don't think it's good for the network for some miners to discover a performance gain that is not disclosed, so I think it's completely unacceptable to hide this fact.

Furthermore, we ought to change the defaults so that the optimal performance would happen by default.  That's outside the scope of this PR, but if there's agreement that this should be done, I'd be happy to implement and open a separate pull.  (The only reason I haven't argued more strenuously for this is that I was satisfied with disclosure in the release notes.)

sipa replied:
> So if ordering by fee-per-bytes is a performance issue, just order by fee-per-weight and close the block out when the size limit is hit. 

That's exactly what the current code does. But it still needs to actually compute the serialized size all the time to know when the size is hit.

> As this stands, it is de facto pushing miners toward a configuration that ought to be discouraged when segwit activates.

I disagree, sorry. Long-term, constraining block weight should be the only limit if we want predictable fees and incentive compatibility with UTXO set size growth. Actual serialized size of blocks on the network shouldn't matter if it's made independent from propagation times by better-than-naive block relay protocols.

luke-jr replied:
I'm talking about right now, not long-term. It is not acceptable for miners to be pushed off blockmaxsize <= 1M for the time being. If that means adding a cache of the real size of the transactions, that sounds like a proper avenue to go down.

laanwj commented:
Thanks for writing this up @sdaftuar . Looks like a clear story, I'm just going to merge this, if anyone has suggested changes they can be made later.

sipa commented:
@luke-jr The default is still max 750 kB (incurring the likely small performance loss), and this release does not have any means of increasing above 1 MB.

luke-jr commented:
But the release notes as currently written give incentive to take blockmaxsize and convert to blockmaxweight which can potentially produce blocks up to 3 MB once segwit is deployed.
